test/Mono.Linker.Tests/TestCasesRunner/ResultChecker.cs

```diff
@@ -663,189 +662,368 @@ void VerifyKeptAllTypesAndMembersInAssembly (AssemblyDefinition linked)
-				switch (attr.AttributeType.Name) {
-
-				case nameof (LogContainsAttribute): {
-						var expectedMessage = (string) attr.ConstructorArguments[0].Value;
-
-						List<MessageContainer> matchedMessages;
-						if ((bool) attr.ConstructorArguments[1].Value)
-							matchedMessages = loggedMessages.Where (m => Regex.IsMatch (m.ToString (), expectedMessage)).ToList ();
-						else
-							matchedMessages = loggedMessages.Where (m => m.ToString ().Contains (expectedMessage)).ToList (); ;
-						Assert.IsTrue (
-							matchedMessages.Count > 0,
-							$"Expected to find logged message matching `{expectedMessage}`, but no such message was found.{Environment.NewLine}Logged messages:{Environment.NewLine}{string.Join (Environment.NewLine, loggedMessages)}");
-
-						foreach (var matchedMessage in matchedMessages)
-							loggedMessages.Remove (matchedMessage);
-					}
-					break;
+			foreach (var attrProvider in GetAttributeProviders (original)) {
+				bool foundReflectionAccessPatternAttributesToVerify = false;
+				foreach (var attr in attrProvider.CustomAttributes) {
+					switch (attr.AttributeType.Name) {
 
-				case nameof (LogDoesNotContainAttribute): {
-						var unexpectedMessage = (string) attr.ConstructorArguments[0].Value;
-						foreach (var loggedMessage in loggedMessages) {
-							Assert.That (() => {
-								if ((bool) attr.ConstructorArguments[1].Value)
-									return !Regex.IsMatch (loggedMessage.ToString (), unexpectedMessage);
-								return !loggedMessage.ToString ().Contains (unexpectedMessage);
-							},
-							$"Expected to not find logged message matching `{unexpectedMessage}`, but found:{Environment.NewLine}{loggedMessage.ToString ()}{Environment.NewLine}Logged messages:{Environment.NewLine}{string.Join (Environment.NewLine, loggedMessages)}");
-						}
-					}
-					break;
+					case nameof (LogContainsAttribute): {
+							var expectedMessage = (string) attr.ConstructorArguments[0].Value;
 
-				case nameof (ExpectedWarningAttribute): {
-						var expectedWarningCode = (string) attr.GetConstructorArgumentValue (0);
-						if (!expectedWarningCode.StartsWith ("IL")) {
-							Assert.Fail ($"The warning code specified in {nameof (ExpectedWarningAttribute)} must start with the 'IL' prefix. Specified value: '{expectedWarningCode}'.");
+							List<MessageContainer> matchedMessages;
+							if ((bool) attr.ConstructorArguments[1].Value)
+								matchedMessages = loggedMessages.Where (m => Regex.IsMatch (m.ToString (), expectedMessage)).ToList ();
+							else
+								matchedMessages = loggedMessages.Where (m => m.ToString ().Contains (expectedMessage)).ToList (); ;
+							Assert.IsTrue (
+								matchedMessages.Count > 0,
+								$"Expected to find logged message matching `{expectedMessage}`, but no such message was found.{Environment.NewLine}Logged messages:{Environment.NewLine}{string.Join (Environment.NewLine, loggedMessages)}");
+
+							loggedMessages.Remove (matchedMessages.First ());
```

Comment:
Just to make sure - it was intentionall to require multiple attributes if the same message appears more than once, right?